### PR TITLE
student feedback results view: rubric stats: show stats for 'received' responses #3744

### LINKED ISSUE #3744: student feedback results view: rubric stats: show stats for 'received' responses
currently shows stats for both received and given responses together. Students may be more interested in the stats of 'received' responses rather than all visible responses.

### PR DESCRIPTION
**Pull request by leeyimin**
Fixes #3744

When showing feedback results of rubric question to students, the responses shown in the summary table is filtered to contain responses with the student as the recipient (either as a individual or team depending on question recipient type) if such responses exist. Else, all visible responses are shown in summary table.

### CODE REVIEW

**Comment by leeyimin** [comment]
Summary of test case and expected output

**Second Session**
Q1: To team members including self, visible to all students
- alice -> alice : No, No response
- alice -> benny: No, Yes
- benny -> alice: Yes, No
- benny -> benny: Yes, Yes
- colin -> colin: No, Yes

**Fourth Session**
Q1: To none, visible to all students
Q2: To team members including self, anonymous to recipient
Q3: To other teams, anonymous to recipient
For all questions,
- alice: Yes, Yes
- benny: No, No
- colin: Yes, No

Student feedback submit page (Second session):
image

Student feedback result page (Second session):
image

Student feedback result extended (Fourth session):
image
image
image

Instructors feedback result (Second session):
Question view
image
Giver Recipient Question View
image
image
image
Giver Question Recipient View
image
image
Recipient Question Giver View
image
image
Recipient Giver Question View
image
image
image

**Comment by JosephineKwa** [comment]
Hey @leeyimin just checking with you 😮, is this ready for review?

**Comment by leeyimin** [comment]
@JosephineKwa Yes. Ready for review!

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java [hunk 1]
```diff
@@ -82,14 +89,14 @@ private void testInstructorResultsPage() throws Exception {
         instructorResultsPage.waitForPanelsToExpand();
         instructorResultsPage.verifyHtmlMainContent("/instructorFeedbackResultsPageRubricGQRView.html");
         
-        // Recipient Giver Question View
+        // Recipient Question Giver View
```
Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: Nice catch 😛

**Comment by leeyimin** [comment]
@JosephineKwa Changes made!

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 2]
```diff
@@ -593,14 +594,74 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+        List<FeedbackResponseAttributes> receivedResponses = new ArrayList<>();
+        String recipientString = toFilterByTeams ? bundle.getTeamNameForEmail(studentEmail) : studentEmail;
+
+        for (FeedbackResponseAttributes response : responses) {
+            boolean isReceivedResponse = response.recipient.equals(recipientString);
+            if (isReceivedResponse) {
+                receivedResponses.add(response);
+            }
+        }
+        if (receivedResponses.isEmpty()) {
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: @JosephineKwa I need a little more details about this question.

Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: For rubric questions where students are recipients, if a student receives responses he is shown only the stats of responses that he received, otherwise if he did not receive any response, he will be shown the stats of all visible responses. I was wondering if this behaviour was alright, or should we not show the stats of visible responses if the student did not receive any responses. 

(This to applies to teams as well)

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: @JosephineKwa @leeyimin I think we should not use one as the substitute for the other as that can be misleading (some instructors might not want students to see overall stats anyway). If no responses received, we should keep it empty or show `N/A` or something.
We should also show 'average excluding self responses' but that can be a separate issue.

**Comment by leeyimin** [comment]
@JosephineKwa Removed the behaviour of showing visible responses when there are no received responses. Ready for review.

No received response: 
image

**Comment by damithc** [comment]
Showing `0.0` as average when there are no responses is not exactly correct. 0 is a legit average value which should not be mixed with 'no responses' situation.

**Comment by leeyimin** [comment]
@damithc @JosephineKwa "-" is used to replace percentage frequency and average when there are no responses. However,  the number of responses in the brackets are still shown to be consistent with the format for subquestions with responses. 


image

**Comment by damithc** [comment]
> @damithc @JosephineKwa "-" is used to replace percentage frequency and average when there are no responses. However, the number of responses in the brackets are still shown to be consistent with the format for subquestions with responses.

UI looks fine @leeyimin

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 3]
```diff
@@ -525,14 +526,17 @@ public String getQuestionAdditionalInfoHtml(int questionNumber, String additiona
...
-            String studentEmail,
-            FeedbackSessionResultsBundle bundle,
-            String view) {
+            FeedbackQuestionAttributes question, String studentEmail,
+            FeedbackSessionResultsBundle bundle, String view) {
+
+        List<FeedbackResponseAttributes> responsesForStatistics =
+                filterResponsesForStatistics(responses, question, studentEmail, bundle, view);
+
+        FeedbackRubricQuestionDetails feedbackQuestionDetails =
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Again, unless it is necessary for this issue, do avoid making unrelated edits. Open a separate issue if you feel that the naming can be improved

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 4]
```diff
@@ -567,79 +569,160 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
     private float[][] calculateRubricStats(List<FeedbackResponseAttributes> responses,
-            FeedbackQuestionAttributes question) {
-        FeedbackRubricQuestionDetails fqd = (FeedbackRubricQuestionDetails) question.getQuestionDetails();
-        
-        // Initialize response frequency variable, used to store frequency each choice is selected.
-        int[][] responseFrequency = calculateResponseFrequency(responses, fqd);
-        
-        return getPercentageFrequencyAndAverage(responseFrequency, fqd);
-    }
+                                           FeedbackRubricQuestionDetails feedbackQuestionDetails) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Is it necessary for these refactorings to be in this PR? Would be better to handle such changes in a separate issue

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java (hunk 1 above)
Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: My apologies, I should have mentioned that it was best to avoid unrelated changes 😓. I thought I'd leave it as you had already made the change.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java (hunk 4 above)
Comment (leeyimin) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: I modified the behaviour of the two methods and found the previous code quite hard to read so I thought it would be better to refactor it in this PR too.

File: src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java (hunk 1 above)
Comment (leeyimin) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: Should I revert the change and open a new issue then?

Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/browsertests/FeedbackRubricQuestionUiTest.java]: Nope not necessary! Just wanted to mention it here for future reference

**Comment by leeyimin** [comment]
@JosephineKwa Thanks and sorry about the unrelated changes! Ready for review.

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 5]
```diff
@@ -567,79 +571,160 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
     }
 
     /**
      * Calculates the response frequency for each choice
+     * Last element in the row stores the total number of responses for the sub-question.
      */
     private int[][] calculateResponseFrequency(List<FeedbackResponseAttributes> responses,
                                                FeedbackRubricQuestionDetails fqd) {
         int[][] responseFrequency = new int[fqd.numOfRubricSubQuestions][];
+        int totalNumberPosition = fqd.numOfRubricChoices;
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Maybe `responseTotalIndex`?

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 6]
```diff
@@ -567,79 +571,160 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+        // calculate percentage frequencies and average value
+        for (int i = 0; i < percentageFrequencyAndAverageValue.length; i++) {
+            int totalForSubQuestion = responseFrequency[i][fqd.numOfRubricChoices];
+            //continue to next row if no response for this sub-question
+            if (totalForSubQuestion == 0) {
+                continue;
+            }
+            // Divide responsesFrequency by totalForSubQuestion to get percentage
+            for (int j = 0; j < numOfRubricChoices; j++) {
+                percentageFrequencyAndAverageValue[i][j] = (float) responseFrequency[i][j] / totalForSubQuestion;
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Move `(float) responseFrequency[i][j] / totalForSubQuestion;` to a new line

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 7]
```diff
@@ -567,79 +571,160 @@ public String getQuestionResultStatisticsHtml(List<FeedbackResponseAttributes> r
...
+            }
+            // Calculate the average for each sub-question
+            if (fqd.hasAssignedWeights) {
+                for (int j = 0; j < numOfRubricChoices; j++) {
+                    float choiceWeight = (float)
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Move `(float)` to the next line

**Comment by leeyimin** [comment]
@JosephineKwa Ready for review!

File: src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java [hunk 8]
```diff
@@ -746,18 +770,24 @@ public String getQuestionResultStatisticsCsv(
         DecimalFormat dfAverage = new DecimalFormat("0.00");
 
         int[][] responseFrequency = calculateResponseFrequency(responses, this);
-        float[][] rubricStats = calculateRubricStats(responses, question);
-        
+        float[][] rubricStats = calculateRubricStats(responses, this);
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: Hey perhaps it would also be good to add `responseFrequency` as a param to `calculateRubricStats`, to avoid having to calculate it again. You could do in in a separate PR if you'd like.

Comment (leeyimin) [inline src/main/java/teammates/common/datatransfer/FeedbackRubricQuestionDetails.java]: I did that at first but I realized it would make the dependency of calculateRubricsStats on calculateResponseFrequency unclear. I will open a separate PR to deal with that.

**Comment by leeyimin** [comment]
@JosephineKwa Ready for review
